File: sdk-ts/src/crypto.ts
/**
 * Cryptographic primitives: ed25519 + blake3
 */

import * as ed from '@noble/ed25519';
import { blake3 } from '@noble/hashes/blake3';
import { bytesToHex, hexToBytes } from '@noble/hashes/utils';

export function hash(data: Uint8Array): string {
  return bytesToHex(blake3(data));
}

export function hashHex(hexData: string): string {
  return hash(hexToBytes(hexData));
}

export async function generateKeypair(): Promise<{ privateKey: string; publicKey: string }> {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = await ed.getPublicKeyAsync(privateKey);
  return {
    privateKey: bytesToHex(privateKey),
    publicKey: bytesToHex(publicKey)
  };
}

export async function sign(messageHash: string, privateKeyHex: string): Promise<string> {
  const messageBytes = hexToBytes(messageHash);
  const privateKeyBytes = hexToBytes(privateKeyHex);
  const signature = await ed.signAsync(messageBytes, privateKeyBytes);
  return bytesToHex(signature);
}

export async function verify(
  messageHash: string,
  signatureHex: string,
  publicKeyHex: string
): Promise<boolean> {
  try {
    const messageBytes = hexToBytes(messageHash);
    const signatureBytes = hexToBytes(signatureHex);
    const publicKeyBytes = hexToBytes(publicKeyHex);
    return await ed.verifyAsync(signatureBytes, messageBytes, publicKeyBytes);
  } catch {
    return false;
  }
}

export { bytesToHex, hexToBytes };

File: sdk-py/primordia_sdk/crypto.py
"""
Cryptographic primitives: ed25519 + blake3
"""

import blake3
from nacl.signing import SigningKey, VerifyKey
from nacl.exceptions import BadSignatureError
from typing import Tuple


def hash(data: bytes) -> str:
    """Compute blake3 hash of data, return hex string."""
    return blake3.blake3(data).hexdigest()


def hash_bytes(data: bytes) -> str:
    """Deprecated: use hash() instead. Compute blake3 hash of data, return hex string."""
    return hash(data)


def generate_keypair() -> Tuple[str, str]:
    """Generate ed25519 keypair, return (private_key_hex, public_key_hex)."""
    signing_key = SigningKey.generate()
    private_key = signing_key.encode().hex()
    public_key = signing_key.verify_key.encode().hex()
    return private_key, public_key


def sign(message_hash: str, private_key_hex: str) -> str:
    """Sign a message hash with ed25519 private key, return signature hex."""
    message_bytes = bytes.fromhex(message_hash)
    private_key_bytes = bytes.fromhex(private_key_hex)
    signing_key = SigningKey(private_key_bytes)
    signed = signing_key.sign(message_bytes)
    # PyNaCl returns message + signature, we just want signature
    signature = signed.signature
    return signature.hex()


def verify(message_hash: str, signature_hex: str, public_key_hex: str) -> bool:
    """Verify ed25519 signature, return True if valid."""
    try:
        message_bytes = bytes.fromhex(message_hash)
        signature_bytes = bytes.fromhex(signature_hex)
        public_key_bytes = bytes.fromhex(public_key_hex)
        verify_key = VerifyKey(public_key_bytes)
        verify_key.verify(message_bytes, signature_bytes)
        return True
    except (BadSignatureError, ValueError, Exception):
        return False
